executor/executor.go

```diff
@@ -0,0 +1,235 @@
+	return &Executor{
+		Command:       cmd,
+		LogInterval:   1 * time.Minute,
+		TerminateChan: make(chan bool, 1),
+		LogFunc:       logrus.Debugf,
+		stdout:        nil,
+		stderr:        nil,
+	}
+}
+
+func (e *Executor) String() string {
+	return fmt.Sprintf("%v (%v) (pid: %v)", e.Command.Args, e.Command.Path, e.PID())
+}
+
+// Start starts the command in the Executor context. It returns any error upon
+// starting the process, but does not wait for it to complete. You may control
+// it in a variety of ways (see Executor for more information).
+func (e *Executor) Start() error {
+	e.Command.Stdin = e.Stdin
+
+	e.startTime = time.Now()
+	e.terminateLogger = make(chan struct{}, 1)
+
+	var err error
+
+	e.stdout, err = e.Command.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	e.stderr, err = e.Command.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	if err := e.Command.Start(); err != nil {
+		e.LogFunc("Error executing %v: %v", e, err)
+		return err
+	}
+
+	go e.waitForStop()
+
+	if e.Timeout != 0 {
+		go e.terminateTimeout()
+	}
+
+	go e.logInterval()
+
+	return nil
+}
+
+// TimeRunning returns the amount of time the program is or was running. Also
+// see ExecResult.Runtime.
+func (e *Executor) TimeRunning() time.Duration {
+	return time.Now().Sub(e.startTime)
+}
+
+func (e *Executor) terminateTimeout() {
+	time.Sleep(e.Timeout)
+	e.TerminateChan <- true
+}
+
+func (e *Executor) waitForStop() {
+	terminate := <-e.TerminateChan
+
+	if terminate {
+		if e.Command.Process == nil {
+			e.LogFunc("Could not terminate non-running command %v", e)
+		} else {
+			e.LogFunc("Command %v terminated due to timeout. It may not have finished!", e)
+			e.Command.Process.Kill()
+		}
+	}
+
+	e.terminateLogger <- struct{}{}
+	return
+}
+
+func (e *Executor) logInterval() {
+	for {
+		time.Sleep(e.LogInterval)
+		select {
+		case <-e.terminateLogger:
+			return
+		default:
+			e.LogFunc("%v has been running for %v", e, e.TimeRunning())
+		}
+	}
+}
+
+// PID yields the pid of the process (dead or alive), or 0 if the process has
+// not been run yet.
+func (e *Executor) PID() uint32 {
+	if e.Command.Process != nil {
+		return uint32(e.Command.Process.Pid)
+	}
+
+	return 0
+}
+
+// Stop waits for the process and return an ExecResult. The program died due to
+// another problem without returning an exit status, a nil result is yielded.B
+//
+// If the ExecResult is returned, error will be nil, regardless of the
+// ExitError returned by (*exec.Cmd).Wait(). If an error is returned it should
+// be handled appropriately.
+func (e *Executor) Stop() (*ExecResult, error) {
+	stdout, err := ioutil.ReadAll(e.stdout)
+	if err != nil {
+		return nil, err
+	}
+
+	stderr, err := ioutil.ReadAll(e.stderr)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &ExecResult{
+		executor: e,
+		Stdout:   string(stdout),
+		Stderr:   string(stderr),
+	}
+
+	err = e.Command.Wait()
+
+	e.TerminateChan <- false // signal goroutines to terminate gracefully
```

Comment:
may be we should `defer` this in the beginning itself otherwise if we return early we may leave behind some go routines?